funnel/utils/markdown/tabs.py

```diff
@@ -0,0 +1,245 @@
+"""MDIT renderer and helpers for tabs."""
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from functools import reduce
+from typing import ClassVar, Dict, List, Optional, Tuple
+
+__all__ = ['render_tab']
+
+
+def get_tab_tokens(tokens):
+    return [
+        {
+            'index': i,
+            'nesting': token.nesting,
+            'info': token.info,
+            'key': '-'.join([token.markup, str(token.level)]),
+        }
+        for i, token in enumerate(tokens)
+        if token.type in ('container_tab_open', 'container_tab_close')
+    ]
+
+
+def get_pair(tab_tokens, start=0) -> Optional[Tuple[int, int]]:
+    i = 1
+    while i < len(tab_tokens):
+        if (
+            tab_tokens[i]['nesting'] == -1
+            and tab_tokens[0]['key'] == tab_tokens[i]['key']
+        ):
+            break
+        i += 1
+    if i >= len(tab_tokens):
+        return None
+    return (start, start + i)
+
+
+def render_tab(self, tokens, idx, _options, env):
+    if 'manager' not in env:
+        env['manager'] = TabsManager(get_tab_tokens(tokens))
```

Comment:
Another of our coding conventions it to have classes first in the file, functions second. The full sequence:

1. imports
2. exceptions
3. constants and enums
4. classes
5. functions